clis/teliod/src/command_listener.rs

```diff
@@ -92,40 +69,69 @@ impl CommandListener {
                 // wait for a response from telio runner
                 if let Some(response) = self.telio_task_response_rx.recv().await {
                     // send the response to teliod client
-                    self.send_response(response).await?;
+                    connection.respond(response.serialize()).await?;
                 }
             }
-            Err(TeliodError::InvalidCommand(cmd)) => {
-                warn!("Received invalid command from client: {}", cmd);
-                eprintln!("Received invalid command from client: {}", cmd);
-            }
-            Err(error) => {
-                warn!("CommandListener error {}", error);
-                return Err(error);
-            }
         }
         Ok(())
     }
+
+    pub async fn handle_client_connection(&mut self) -> Result<ClientCmd, TeliodError> {
+        let mut connection = self.socket.accept().await?;
+        let command_str = connection.read_command().await?;
+
+        if let Ok(command) = serde_json::from_str::<ClientCmd>(&command_str) {
+            self.process_command(&command, &mut connection).await?;
+            return Ok(command);
+        } else {
+            error!("Received invalid command from client: {}", command_str);
+            connection
+                .respond(
+                    CommandResponse::Err(format!("Invalid command: {}", command_str)).serialize(),
+                )
+                .await?;
+            return Err(TeliodError::InvalidCommand(command_str));
+        }
+    }
 }
 
 #[cfg(test)]
 mod tests {
     use super::*;
     use crate::{ClientCmd, CommandResponse, DaemonSocket};
+    use serial_test::serial;
     use std::path::Path;
-    use tokio::{io::AsyncWriteExt, net::UnixStream, sync::mpsc};
+    use tokio::{
+        io::{AsyncReadExt, AsyncWriteExt},
+        net::UnixStream,
+        sync::mpsc,
+    };
 
     const TEST_SOCKET_PATH: &str = "test_socket";
 
+    // Helper to create a fake command listener
     fn make_command_listener() -> CommandListener {
         let (tx, _rx) = mpsc::channel(1);
         let (_tx, rx) = mpsc::channel(1);
 
         let socket = DaemonSocket::new(Path::new(TEST_SOCKET_PATH)).unwrap();
+
         CommandListener::new(socket, tx.clone(), rx)
     }
 
-    async fn client_send_command(cmd: &str) -> std::io::Result<()> {
+    // Simulate client sending command and waiting for response
+    async fn client_send_command(cmd: &str) -> Result<CommandResponse, std::io::Error> {
```

Comment:
Nit: Isn't it just `std::io::Result<CommandResponse>`?

Reply:
It is indeed 😅

Reply:
Fixed